container/array/array.go

```diff
@@ -0,0 +1,164 @@
+// The returned pointer will be zero-initialized.
+func (a *Array[T]) Next() (ptr *T, index int) {
+	return a.NextWithPool(nil)
+}
+
+// NextWithPool obtains a pointer to the next slot, allocating new space from the
+// optionally provided pool if needed. The returned pointer will be zero-initialized.
+func (a *Array[T]) NextWithPool(pool *Pool[T]) (ptr *T, index int) {
+	i := len(a.blocks) - 1
+	if i >= 0 && a.blocks[i].available() {
+		ptr, index = a.blocks[i].next()
+	} else {
+		a.blocks = append(a.blocks, block[T]{slots: pool.get(), count: 1})
+		i++
+		ptr = &a.blocks[i].slots[index]
+	}
+	index += i * BlockLen[T]()
+	return
+}
+
+// Pop removes the last item from the array and returns the value.
+func (a *Array[T]) Pop() T {
+	return a.PopWithPool(nil)
+}
+
+// PopWithPool removes the last item from the array. If the popped value is the
+// last in a block, the block will be removed and returned to the pool.
+func (a *Array[T]) PopWithPool(pool *Pool[T]) T {
+	i := len(a.blocks) - 1
+	v := a.blocks[i].pop()
+
+	if a.blocks[i].len() == 0 {
+		free := a.blocks[i].reset()
+		if pool != nil {
+			pool.put(free)
+		}
+		a.blocks = a.blocks[:i]
+
+		if len(a.blocks) < cap(a.blocks)/4 {
+			blocks := make([]block[T], len(a.blocks), cap(a.blocks)/2)
+			copy(blocks, a.blocks)
+			a.blocks = blocks
+		}
+	}
+	return v
+}
+
+// ShiftBlocks removes `n` number of blocks (`n * BlockLen[T]()` entries) from
+// the beginning of the array, shifting the remaining blocks down. If `n` exceeds
+// the number of blocks in the array, none are removed and false is returned.
+func (a *Array[T]) ShiftBlocks(n int) bool {
+	return a.ShiftBlocksWithPool(n, nil)
+}
+
+// ShiftBlocksWithPool removes whole blocks from the array like ShiftBlocks. If
+// a pool is provided, the removed blocks will be given back to it.
+//
+// When using a pool, it is important to ensure any entries that contain heap
+// allocations are cleared. For example, the entire blocks may be cleared at once:
+//
+//	type Item { values []int }
+//	func removeBlocks(items array.Pool[Item], blocks int, pool *array.Pool[Item]) {
+//		n := array.BlockLen[Item] * blocks
+//		for i := 0; i < n; i++ {
+//			items.Ptr(i).values = nil
+//		}
+//		items.ShiftBlocksWithPool(blocks, &pool)
+//	}
+func (a *Array[T]) ShiftBlocksWithPool(n int, pool *Pool[T]) bool {
+	if n >= len(a.blocks) {
+		return false
+	}
+
+	if pool != nil {
+		for i := 0; i < n; i++ {
+			pool.put(a.blocks[i].slots)
+		}
+	}
+	a.blocks = a.blocks[n:]
+
+	return true
+}
+
+// Get obtains the value of an entry at a specific offset. Like a Go slice
+// this results in a panic if the value is outside the array bounds.
+func (a *Array[T]) Get(i int) T {
+	return *a.Ptr(i)
+}
+
+// Ptr obtains the pointer to a value for an entry at a specific offset. The
+// returned pointer is guaranteed not to change until the index is popped from
+// the array. Like a Go slice this results in a panic if the value is outside
+// the array bounds.
```

Comment:
If you did get a panic would you be able to adequately diagnose the problem?

Reply:
I can make the panic message a bit more explicit